### PR TITLE
Add logical key data to KeyboardInput

### PR DESCRIPTION
**Pull request by Vrixyz**
Add logical key data to KeyboardInput

Addresses an item of https://github.com/bevyengine/bevy/issues/11052

### CODE REVIEW

**Comment by rparrett** [comment]
It seems like this would close #11297 for me. At least the behavior between web/native on MacOS for `KeyInput::logical_key(Key::Character(s))` seems predictable enough to work with. It would be great if someone could test with a non-US keyboard.

I am still sort of baffled by the myriad of ways of getting a character from a winit key event, so I hope I'm doing this correctly.

I'm testing in this branch of `bevy_simple_text_input`: https://github.com/rparrett/bevy_simple_text_input/blob/bevy_main/src/lib.rs#L75

File: crates/bevy_winit/src/converters.rs [hunk 1]
```diff
@@ -283,6 +285,348 @@ pub fn convert_physical_key_code(virtual_key_code: winit::keyboard::PhysicalKey)
     }
 }
 
+pub fn convert_logical_key(logical_key_code: &winit::keyboard::Key) -> bevy_input::keyboard::Key {
```
Comment (alice-i-cecile) [inline crates/bevy_winit/src/converters.rs:288]: This feels like it should be a `From` impl?

File: crates/bevy_input/src/keyboard.rs [hunk 2]
```diff
@@ -650,3 +653,795 @@ pub enum KeyCode {
...
+/// It is used as the generic `T` value of an [`Input`] to create a `Res<Input<Key>>`.
+/// The resource values are mapped to the current layout of the keyboard and correlate to a [`KeyCode`].
+///
+/// ## Updating
+///
+/// The resource is updated inside of the [`keyboard_input_system`].
+///
+/// ## Technical
+///
+/// Its values map 1 to 1 to winit's Key.
```
Comment (alice-i-cecile) [inline crates/bevy_input/src/keyboard.rs:698]: Doc link here would be useful.

File: crates/bevy_input/src/keyboard.rs [hunk 3]
```diff
@@ -650,3 +653,795 @@ pub enum KeyCode {
...
 }
+
+/// Contains the platform-native logical key identifier
+///
+/// Exactly what that means differs from platform to platform, but the values are to some degree
+/// tied to the currently active keyboard layout. The same key on the same keyboard may also report
+/// different values on different platforms, which is one of the reasons this is a per-platform
+/// enum.
+///
+/// This enum is primarily used to store raw keysym when Winit doesn't map a given native logical
```
Comment (alice-i-cecile) [inline crates/bevy_input/src/keyboard.rs:664]: keysym?

**Review by alice-i-cecile — APPROVED** [review]
A couple of doc suggestions but nothing blocking.

**Comment by alice-i-cecile** [comment]
> I am still sort of baffled by the myriad of ways of getting a character from a winit key event, so I hope I'm doing this correctly.

I would *love* to see an example for this, but it probably shouldn't be this PR.

File: crates/bevy_input/src/keyboard.rs (hunk 3 above)
Comment (matiqo15) [reply crates/bevy_input/src/keyboard.rs:664]: I believe it's short for key symbol.

File: crates/bevy_input/src/keyboard.rs [hunk 4]
```diff
@@ -650,3 +653,795 @@ pub enum KeyCode {
...
+    /// (`KEYCODE_BRIGHTNESS_DOWN`)
+    BrightnessDown,
+    /// The Brightness Up key. Typically controls the display brightness. (`KEYCODE_BRIGHTNESS_UP`)
+    BrightnessUp,
+    /// Toggle removable media to eject (open) and insert (close) state. (`KEYCODE_MEDIA_EJECT`)
+    Eject,
+    /// LogOff
+    LogOff,
+    /// Toggle power state. (`KEYCODE_POWER`)
+    /// Note: Note: Some devices might not expose this key to the operating environment.
```
Comment (matiqo15) [inline crates/bevy_input/src/keyboard.rs:878]:
```suggestion
    /// Note: Some devices might not expose this key to the operating environment.
```

File: crates/bevy_input/src/keyboard.rs [hunk 5]
```diff
@@ -650,3 +653,795 @@ pub enum KeyCode {
...
+    ColorF4Grey,
+    /// General purpose color-coded media function key, as index 5 (brown). (`VK_COLORED_KEY_5`)
+    ColorF5Brown,
+    /// Toggle the display of Closed Captions. (`VK_CC`, `KEYCODE_CAPTIONS`)
+    ClosedCaptionToggle,
+    /// Adjust brightness of device, by toggling between or cycling through states. (`VK_DIMMER`)
+    Dimmer,
+    /// Swap video sources. (`VK_DISPLAY_SWAP`)
+    DisplaySwap,
+    /// Select Digital Video Rrecorder. (`KEYCODE_DVR`)
```
Comment (matiqo15) [inline crates/bevy_input/src/keyboard.rs:1224]:
```suggestion
    /// Select Digital Video Recorder. (`KEYCODE_DVR`)
```

**Review by matiqo15 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@Vrixyz lemme know when you're ready and then I'll merge. Otherwise, I'll merge Monday.

File: crates/bevy_input/src/keyboard.rs (hunk 3 above)
Comment (Vrixyz) [reply crates/bevy_input/src/keyboard.rs:664]: Those descriptions are coming mostly from winit, I'm not sure referring to them as keysym helps much the user indeed. A google search on that term doesn't respond with a clear description, so I assume it's a vague term for referring to key codes, or logical key code names.

I'm not opposed to remove the term. Or maybe bikeshed that in another issue

Comment (Vrixyz) [reply crates/bevy_input/src/keyboard.rs:664]: @alice-i-cecile another question on that line is the reference to winit which we should probably avoid within bevy_input in favor of "window backend" or equivalent.

Comment (alice-i-cecile) [reply crates/bevy_input/src/keyboard.rs:664]: Yeah, "windowing backend" is slightly better.

File: crates/bevy_input/src/keyboard.rs (hunk 2 above)
Comment (Vrixyz) [reply crates/bevy_input/src/keyboard.rs:698]: We can't really "doc link", not to the exact type, because we could rely on another backend. This note is more for contributors to understand where this is coming from. I think a link to winit would be useful but maybe difficult to maintain ?

File: crates/bevy_input/src/keyboard.rs (hunk 3 above)
Comment (Vrixyz) [reply crates/bevy_input/src/keyboard.rs:664]: at the same time, I'm curious how many users don't use winit ; blurring the line of the dependency to winit comes at a price of increased complexity: the user has to understand what is the "windowing backend".

It can be argued it's a "small complexity", or an "implementation detail", but I hate when I have to jump through a lot of "soft references", I'll consider this "extreme abstraction" shadowing winit controversial.

It's listed in the todo on https://github.com/bevyengine/bevy/issues/11052 so this discussion will come again :)

File: crates/bevy_input/src/keyboard.rs [hunk 6]
```diff
@@ -650,3 +653,795 @@ pub enum KeyCode {
...
+    /// A "key value string".
+    Web(SmolStr),
+}
+
+/// The logical key code of a [`KeyboardInput`].
+///
+/// ## Usage
+///
+/// It is used as the generic `T` value of an [`Input`] to create a `Res<Input<Key>>`.
+/// The resource values are mapped to the current layout of the keyboard and correlate to a [`KeyCode`].
```
Comment (Vrixyz) [inline crates/bevy_input/src/keyboard.rs:690]: This is wrong, it's only used in `KeyboardInput`.

I wanted to support that use case but decided to remove it there: https://github.com/bevyengine/bevy/pull/10702/commits/9c7a452cc3c9547049546a6da2572806e3235f7d#diff-175a1b1cfe0a099d4b8927aff78925ab491b7f48c14a0ad99ad1f34c688e3e51.

More insights here: https://github.com/Vrixyz/bevy/pull/3

TL;DR: adding support for reliable `is_released(a key)` after `is_pressed(same key)` has fired is controversial at best.

File: crates/bevy_input/src/keyboard.rs [hunk 7]
```diff
@@ -86,16 +87,18 @@ use bevy_reflect::{ReflectDeserialize, ReflectSerialize};
...
     feature = "serialize",
     derive(serde::Serialize, serde::Deserialize),
     reflect(Serialize, Deserialize)
 )]
 pub struct KeyboardInput {
-    /// The key code of the key.
+    /// The physical key code of the key.
     pub key_code: KeyCode,
+    /// The logical key of the input
+    pub logical_key: Key,
```
Comment (Vrixyz) [inline crates/bevy_input/src/keyboard.rs:101]: maybe `key_logical` ? so nomenclature is similar for similar properties.

And rename `key_code`to `key_physical` ? Probably as a follow up, close to an existing one in https://github.com/bevyengine/bevy/issues/11052 ("Rename KeyCode to PhysicalKeyCode")

Comment (Vrixyz) [reply crates/bevy_input/src/keyboard.rs:101]: or I guess we'll probably rename `key_code` to `physical_key`.

I just like the similar starting keystroks for similar properties, because it's autocompletion friendly, but I won't insist on it. (especially since winit uses `physical_key` and `logical_key`, so LGTM for now.

File: crates/bevy_input/src/lib.rs [hunk 8]
```diff
@@ -114,7 +114,10 @@ impl Plugin for InputPlugin {
 
         // Register keyboard types
         app.register_type::<KeyboardInput>()
-            .register_type::<KeyCode>();
+            .register_type::<KeyCode>()
+            .register_type::<NativeKeyCode>()
```
Comment (Vrixyz) [inline crates/bevy_input/src/lib.rs:118]: That's a tiny bit out of scope, but I guess it's alright.

**Comment by Vrixyz** [comment]
> @Vrixyz lemme know when you're ready and then I'll merge. Otherwise, I'll merge Monday.

Thanks! I think I'm done with it now, I left a few self-reviews to guide more discussion if need be.
